### DISCUSSION TITLE
Megakernel tutorial

### DISCUSSION THREAD

**Discussion by saeedhd96** [mitsuba-renderer/mitsuba3#479] [Q&A]
Hi,

I don't have too much knowledge about how the megakernel works inside mitsuba, and I find it very confusing when and where I should use function calls like `dr.eval(),` `dr.schedule()` and generally things pertaining to efficiency in megakernel. Is there a tutorial or guide for that?

Thanks,
-Saeed

**Comment by njroussel** [comment]
Hi @saeedhd96 

I don't know how this slipped through my notifications :slightly_frowning_face: 
Unfortunately, we don't have any in-depth guides on megakernels.

For performance, the rule of thumb is to try and keep everything into a single kernel. This [section of the Dr.Jit documentation ](https://drjit.readthedocs.io/en/latest/firststeps-py.html#kernel-launches-caching ) is worth reading and understanding in detail. These logs (enabled with `dr.set_log_level(dr.LogLevel.Info))`) are the best way to understand what kernels are launched and when they are launched.
For example, in an optimization loop, you'll want to be sure that you're launching the same kernels at every iteration. This will save you some backend compilation time because of the cache lookups. You'll also want to check the number of kernels that are launched and why they are potentially many. The kernels sizes are a good hint of what sort of operation is taking place in the kernel.

`dr.schedule()` and `dr.eval()` become useful when you have multiple kernels that are being launched when you actually only want/need one. The [documentation for `dr.schedule()`](https://drjit.readthedocs.io/en/latest/reference.html) has a nice example.
Of course, depending on what you're doing and especially how you're trying to do it, there just has to be multiple kernel launches.
